test/expect-tests/fiber/fiber_tests.ml

```diff
@@ -607,56 +611,50 @@ let%expect_test "Sequence.parallel_iter is indeed parallel" =
   let test ~iter_function =
-    let rec sequence n =
-      if n = 4 then
-        Fiber.return Fiber.Sequence.Nil
-      else
-        Fiber.return (Fiber.Sequence.Cons (n, sequence (n + 1)))
-    in
     Scheduler.run
       (Fiber.finalize
          ~finally:(fun () ->
            Printf.printf "finalized";
            Fiber.return ())
-         (fun () -> iter_function (sequence 1) ~f:(fun _ -> Fiber.return ())))
+         (fun () -> iter_function (stream 1 3) ~f:(fun _ -> Fiber.return ())))
   in
-  test ~iter_function:Fiber.Sequence.sequential_iter;
+  test ~iter_function:Fiber.Stream.In.sequential_iter;
   [%expect {| finalized |}];
 
-  test ~iter_function:Fiber.Sequence.parallel_iter;
+  test ~iter_function:Fiber.Stream.In.parallel_iter;
   [%expect {| finalized |}]
 
-let rec naive_sequence_parallel_iter (t : _ Fiber.Sequence.t) ~f =
-  t >>= function
-  | Nil -> Fiber.return ()
-  | Cons (x, t) ->
+let rec naive_stream_parallel_iter (t : _ Fiber.Stream.In.t) ~f =
+  Fiber.Stream.In.read t >>= function
+  | None -> Fiber.return ()
+  | Some x ->
     Fiber.fork_and_join_unit
       (fun () -> f x)
-      (fun () -> naive_sequence_parallel_iter t ~f)
+      (fun () -> naive_stream_parallel_iter t ~f)
 
-let%expect_test "Sequence.parallel_iter doesn't leak" =
-  (* Check that a naive [parallel_iter] functions on sequences leaks memory,
-     while [Fiber.Sequence.parallel_iter] does not. To do that, we construct a
-     long sequence and iterate over it. At each iteration, we do a full major GC
+let%expect_test "Stream.parallel_iter doesn't leak" =
+  (* Check that a naive [parallel_iter] functions on streams is leaking memory,
+     while [Fiber.Stream.parallel_iter] does not. To do that, we construct a
+     long stream and iterate over it. At each iteration, we do a full major GC
      and count the number of live words. With the naive implementation, we check
      that this number increases while with the right one we check that this
      number is constant.
 
      This test is carefully crafted to avoid creating new live words as we
-     iterate through the sequence. As a result, the only new live words that can
+     iterate through the stream. As a result, the only new live words that can
      appear are because of the iteration function. *)
   let test ~iter_function ~check =
-    let rec sequence n =
-      (* This yield is to ensure that we don't build the whole sequence upfront,
-         which would cause the number of live words to decrease as we iterate
-         through the sequence. *)
-      let* () = Scheduler.yield () in
-      if n = 0 then
-        Fiber.return Fiber.Sequence.Nil
-      else
-        Fiber.return (Fiber.Sequence.Cons ((), sequence (n - 1)))
+    let stream n =
+      let n = ref n in
+      Fiber.Stream.In.create (fun () ->
+          if !n = 0 then
+            Fiber.return None
+          else (
+            decr n;
+            Fiber.return (Some ())
```

Comment:
This can now be `Fiber.Stream.In.of_list (List.init n ~f:Fun.id)`

Reply:
sounds better indeed, will do

Reply:
hmm, actually for this particular test it doesn't quite work as everytime we consume an element we'd free some memory, and this test is sensitive to that.